### PR TITLE
Add subscription version range when subscription creating

### PR DESCRIPTION
**Pull request by chathuranga95**
Related task: https://github.com/wso2-enterprise/choreo/issues/22171

- `VERSION_RANGE` new field is added to the `AM_SUBSCRIPTION` table.
- When subscription is creating for a new API (version starting with `v`), the `VERSION_RANGE` is added as the major version of the API. Eg. When creating a subscription for API v1.2.3, the `VERSION_RANGE` is saved as `v1`
- Adds unit tests to test the added utils

The `Subscription` object is updated by adding a new attribute `versionRange` which contains the above described version range string.

The responses of the following resources are updated.
1. GET /subscriptions
2. GET /subscriptions/{subscription_uuid}
3. POST /subscriptions
4. PUT /subscriptions

Sample response for the `GET /subscriptions` resource. (some attributes are clipped)
```json
{
    "count": 4,
    "list": [
        {
            "subscriptionId": "5778967f-fe53-4b85-b47e-59adbee338eb",
            "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
            "apiId": "64ac0bc64d850605e92dc7bf",
            "apiInfo": {
                "id": "64ac0bc64d850605e92dc7bf",
                ...
            },
            "applicationInfo": {
                "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
                ...
            },
            "throttlingPolicy": "Unlimited",
            "requestedThrottlingPolicy": null,
            "status": "UNBLOCKED",
            "redirectionParams": null,
            "versionRange": "v1"
        },
        {
            "subscriptionId": "f3ff3753-ed2d-42b6-bac7-cf680254778a",
            "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
            "apiId": "64ac121be4ff3b33d63849bf",
            "apiInfo": {
                "id": "64ac121be4ff3b33d63849bf",
                ...
            },
            "applicationInfo": {
                "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
                ...
            },
            "throttlingPolicy": "Bronze",
            "requestedThrottlingPolicy": "Bronze",
            "status": "UNBLOCKED",
            "redirectionParams": null,
            "versionRange": "v1"
        },
        {
            "subscriptionId": "198797a1-fc58-4dba-adf1-d797e364c6b8",
            "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
            "apiId": "64ac09af4d850605e92dc7bd",
            "apiInfo": {
                "id": "64ac09af4d850605e92dc7bd",
                ...
            },
            "applicationInfo": {
                "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
                ...
            },
            "throttlingPolicy": "Bronze",
            "requestedThrottlingPolicy": "Bronze",
            "status": "UNBLOCKED",
            "redirectionParams": null,
            "versionRange": null
        },
        {
            "subscriptionId": "9ee4f52d-81bf-4e36-ab9b-32a99d661bab",
            "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
            "apiId": "64ae63805b6d7b1b0c4c0efb",
            "apiInfo": {
                "id": "64ae63805b6d7b1b0c4c0efb",
                ...
            },
            "applicationInfo": {
                "applicationId": "a19fb96b-0380-41e0-add7-4e1c4e2f7d3d",
                ...
            },
            "throttlingPolicy": "Bronze",
            "requestedThrottlingPolicy": "Bronze",
            "status": "UNBLOCKED",
            "redirectionParams": null,
            "versionRange": "v2"
        }
    ],
    "pagination": null
}
```

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java [hunk 1]
```diff
@@ -530,7 +530,8 @@ public enum ExceptionCodes implements ErrorHandler {
             "Revision deployment request conflicted with the current deployment state of the revision %s. Please try again later", false),
     INVALID_TENANT_CONFIG(9020001, "Invalid tenant-config found", 400, "Invalid tenant-config found with error %s", false),
     INVALID_API_ID(9020002, "Invalid API ID", 404, "The provided API ID is not found %s", false),
-    INVALID_ENVIRONMENT(9020003, "Invalid Environment", 400, "The provided Environment, %s,  is not valid.", false);
+    INVALID_ENVIRONMENT(9020003, "Invalid Environment", 400, "The provided Environment, %s,  is not valid.", false),
+    INVALID_API_SEMANTIC_VERSION(9020004, "Invalid API semantic version", 400, "The provided API version %s is not a valid semantic version.", false);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/ExceptionCodes.java:534]:
```suggestion
    INVALID_API_SEMANTIC_VERSION(9020004, "Invalid API version", 400, "The provided API version %s is not a valid semantic version.", false);
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 2]
```diff
@@ -5101,12 +5104,16 @@ private int addSubscription(Connection connection, ApiTypeWrapper apiTypeWrapper
...
+        String apiVersion;
+        String majorVersionRange;
 
         //Query to check if this subscription already exists
         String checkDuplicateQuery = SQLConstants.CHECK_EXISTING_SUBSCRIPTION_API_SQL;
         if (!isProduct) {
             identifier = apiTypeWrapper.getApi().getId();
             apiUUID = apiTypeWrapper.getApi().getUuid();
+            String[] contextParts = apiTypeWrapper.getApi().getContext().split("/");
+            apiVersion = contextParts[contextParts.length - 1];
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5116]: Why do we get version from the context? can't we take it from the API object?

Comment (chathuranga95) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5117]: We have access to the API object which is inside of the `apiTypeWrapper`. It doesn't have the `version`. That's why this is taken from the context.
Ref: https://github.com/wso2/carbon-apimgt/blob/master/components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/API.java

Comment (chathuranga95) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: API version is taken from the `identifier`. ignore the above comment

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 3]
```diff
@@ -5159,6 +5169,61 @@ private int addSubscription(Connection connection, ApiTypeWrapper apiTypeWrapper
...
+
+                        if ((APIConstants.SubscriptionStatus.UNBLOCKED.equals(subStatus) ||
+                                APIConstants.SubscriptionStatus.ON_HOLD.equals(subStatus) ||
+                                APIConstants.SubscriptionStatus.REJECTED.equals(subStatus)) &&
+                                APIConstants.SubscriptionCreatedStatus.SUBSCRIBE.equals(subCreationStatus)) {
+
+                            // Throw error saying subscription for the api version range already exists.
+                            log.error(String.format("Subscription already exists for API/API Product %s version " +
+                                            "range %s in Application %s", apiTypeWrapper.getName(), versionRange,
+                                    application.getName()));
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5203]: Don't log an throw. Remove log

Comment (malinthaprasan) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:5203]: Pls fix other places too

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java [hunk 4]
```diff
@@ -765,9 +764,28 @@ public Monetization getMonetizationImplClass() throws APIManagementException {
...
+                }
+            } else {
+                throw new APIManagementException(ExceptionCodes.from(ExceptionCodes.VERSION_RANGE_NOT_ALLOWED));
+            }
+        }
+        return subscriptionVersionRange;
+    }
+
     @Override
-    public SubscriptionResponse addSubscription(ApiTypeWrapper apiTypeWrapper, String userId, Application application)
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIConsumerImpl.java:769]: Let's not pass the null values to method arguments when calling from outside. Its very hard to read whether nulls are accepted. We can have an overidden method which doesn't have the nullable argument and call the 2nd method with passing null.

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/utils/ImportUtils.java [hunk 5]
```diff
@@ -171,10 +171,12 @@ public static List<APIIdentifier> importSubscriptions(Set<ExportedSubscribedAPI>
                         // Add subscription if update flag is not specified
                         // It will throw an error if subscriber already exists
                         if (update == null || !update) {
+                            // Version range is added as null since this flow is not used in Choreo
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/utils/ImportUtils.java:174]:
```suggestion
```

File: components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/utils/ImportUtils.java [hunk 6]
```diff
@@ -171,10 +171,12 @@ public static List<APIIdentifier> importSubscriptions(Set<ExportedSubscribedAPI>
...
                             apiConsumer.addSubscription(apiTypeWrapper, userId, application);
                         } else if (!apiConsumer.isSubscribedToApp(subscribedAPI.getApiId(), userId
                                 , application.getId())) {
                             // on update skip subscriptions that already exists
+                            // Version range is added as null since this flow is not used in Choreo
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.rest.api.store.v1/src/main/java/org/wso2/carbon/apimgt/rest/api/store/v1/utils/ImportUtils.java:179]:
```suggestion
```

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIConsumerImplTest.java [hunk 7]
```diff
@@ -840,12 +842,18 @@ public void testAddSubscription() throws APIManagementException {
         String tenantAwareUsername = "user1@"+SAMPLE_TENANT_DOMAIN_1;
         Mockito.when(MultitenantUtils.getTenantAwareUsername(Mockito.eq("user1"))).thenReturn(tenantAwareUsername);
         Mockito.when(apiMgtDAO.addSubscription(apiTypeWrapper, application, APIConstants.SubscriptionStatus.ON_HOLD,
-                tenantAwareUsername)).thenReturn(1);
+                tenantAwareUsername, null)).thenReturn(1);
```
Comment (malinthaprasan) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIConsumerImplTest.java:845]:
```suggestion
                tenantAwareUsername)).thenReturn(1);
```

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**